tests/unit/dtypes/test_shape.py

```diff
@@ -0,0 +1,217 @@
+#
+# Copyright (c) 2022, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import pytest
+
+import merlin.dtypes as md
+from merlin.dtypes.shape import Dimension, Shape
+
+# Dimension
+
+
+def test_empty_dimension():
+    dim = Dimension()
+    assert dim.min == 0
+    assert dim.max is None
+
+
+def test_min_max_val_dimension():
+    dim = Dimension(2, 3)
+    assert dim.min == 2
+    assert dim.max == 3
+
+
+def test_fixed_min_with_unbounded_max():
+    dim = Dimension(2)
```

Comment:
It may make our API more familiar to new users to make `Dimension(2)` correspond to a fixed dimension of size 2 (min=2, max=2) instead of a ragged dimension with min size of 2

Reply:
I haven't seen any mentions of use-cases motivating the addition of the `min` in the dimension size other than it being there before in the value count property. I guess it's a more generic version of a Dimension in tensorflow for example that is either a fixed size or ragged with unknown min/max value.

Reply:
We may need to give some consideration to what the min/max means in the ragged case. For example the ValueCount operator in NVTabular would result in a schema  with min and max set to the min/max observed in the data it was fit on. However, that doesn't necessarily imply that the model or serving system is limited to those min/max values

Reply:
> It may make our API more familiar to new users to make Dimension(2) correspond to a fixed dimension of size 2 (min=2, max=2) instead of a ragged dimension with min size of 2

We considered this, but it made the code rather difficult to write, so we settled on providing syntactic sugar when creating a `ColumnSchema` with the `dims` property. We expect people to either use that path or the `Shape` constructor, which offers the same functionality, instead of interacting with the Dimension constructor directly. In either of those places, you can supply `2` for a dim, which means a dimension of fixed size like you're envisioning.

Reply:
> I haven't seen any mentions of use-cases motivating the addition of the min in the dimension size other than it being there before in the value count property. I guess it's a more generic version of a Dimension in tensorflow for example that is either a fixed size or ragged with unknown min/max value.

We're not quite there yet, but we should be able to directly validate a schema against a `DictArray`/`TensorTable` and match up the shapes of the data with the requirements in the schema. Allowing a non-zero minimum lets us differentiate between list fields that are allowed to be empty and those that aren't. I'm not sure how useful that will end up being though.

Reply:
> We may need to give some consideration to what the min/max means in the ragged case. For example the ValueCount operator in NVTabular would result in a schema with min and max set to the min/max observed in the data it was fit on. However, that doesn't necessarily imply that the model or serving system is limited to those min/max values

Yeah, this is a good point. What I'm envisioning is that each Merlin component updates the schema and stores the updated version alongside its artifacts, so when we save a model if the model is fine with a column having a minimum size of zero, it should update the schema to say so.

Something like this:
```
Parquet files -> NVT inferred schema -> NVT DAG -> NVT annotated schema -> Parquet files w/ annotated schema

Annotated schema -> dataloader transforms -> dataloader output schema -> model input layer

*model training happens*

Saved model w/ input and output schemas -> Systems op
```

Reply:
> We considered this, but it made the code rather difficult to write, so we settled on providing syntactic sugar when creating a `ColumnSchema` with the `dims` property. [...] you can supply `2` for a dim, which means a dimension of fixed size like you're envisioning.

I'd expect to be able to proivde `None` too as part of this syntactic sugar. For example, `Shape((1, None, 10))` to create a shape with 2 fixed and one ragged dimension. This currently raises a ValueError.

Reply:
Supporting `None` as a dimension to the Shape constructor would help with making it _relatively straightforward to (for example) create a Merlin shape from a Tensorflow shape._